### PR TITLE
Issue #1926 - Add support for ':not' modifier within chained search and whole-system search

### PR DESCRIPTION
**Pull request by tbieste**
Issue #1926 - Adds support for the ':not' modifier to be used in chained search and whole-system search.

Signed-off-by: Troy Biesterfeld tbieste@us.ibm.com

### CODE REVIEW

**Comment by tbieste** [comment]
Example SQL queries:

**Patient?_tag:not=tag**
SELECT COUNT(DISTINCT R.LOGICAL_RESOURCE_ID)  FROM Patient_LOGICAL_RESOURCES LR  JOIN Patient_RESOURCES R ON R.RESOURCE_ID = LR.CURRENT_RESOURCE_ID  AND R.IS_DELETED = 'N'
WHERE NOT EXISTS (SELECT 1 FROM Patient_TOKEN_VALUES_V  AS param0 WHERE (param0.PARAMETER_NAME_ID = 1008 AND ((param0.TOKEN_VALUE = ?))) AND LR.LOGICAL_RESOURCE_ID = param0.LOGICAL_RESOURCE_ID)

**?_type=Patient&_tag:not=tag**
SELECT SUM(CNT)  FROM ( SELECT COUNT(DISTINCT LR.LOGICAL_RESOURCE_ID) AS CNT  FROM (SELECT DISTINCT LR.LOGICAL_RESOURCE_ID, LR.LOGICAL_ID, LR.CURRENT_RESOURCE_ID FROM Patient_LOGICAL_RESOURCES LR
WHERE NOT EXISTS (SELECT 1 FROM Patient_TOKEN_VALUES_V  AS param0 WHERE (param0.PARAMETER_NAME_ID = 1008 AND ((param0.TOKEN_VALUE = ?))) AND LR.LOGICAL_RESOURCE_ID = param0.LOGICAL_RESOURCE_ID)) AS LR  JOIN Patient_RESOURCES AS R  ON      R.LOGICAL_RESOURCE_ID = LR.LOGICAL_RESOURCE_ID  AND R.RESOURCE_ID = LR.CURRENT_RESOURCE_ID  AND R.IS_DELETED = 'N') COMBINED_RESULTS

**Observation?subject:Patient._tag:not=tag**
SELECT COUNT(DISTINCT R.LOGICAL_RESOURCE_ID)  FROM Observation_LOGICAL_RESOURCES LR  JOIN Observation_RESOURCES R ON R.RESOURCE_ID = LR.CURRENT_RESOURCE_ID  AND R.IS_DELETED = 'N'  JOIN Observation_TOKEN_VALUES_V  AS param0 ON (param0.PARAMETER_NAME_ID = 1396 AND param0.CODE_SYSTEM_ID = 20005 AND (param0.TOKEN_VALUE IN (SELECT CLR1.LOGICAL_ID FROM Patient_RESOURCES CR1 , Patient_LOGICAL_RESOURCES CLR1 , Patient_TOKEN_VALUES_V CP1 WHERE CR1.RESOURCE_ID = CLR1.CURRENT_RESOURCE_ID AND CR1.IS_DELETED = 'N' AND CP1.LOGICAL_RESOURCE_ID = CR1.LOGICAL_RESOURCE_ID AND  NOT  EXISTS (SELECT 1 FROM Patient_TOKEN_VALUES_V  WHERE (Patient_TOKEN_VALUES_V .PARAMETER_NAME_ID = 1008 AND ((Patient_TOKEN_VALUES_V .TOKEN_VALUE = ?))) AND Patient_TOKEN_VALUES_V .LOGICAL_RESOURCE_ID = CLR1.LOGICAL_RESOURCE_ID)))) AND LR.LOGICAL_RESOURCE_ID = param0.LOGICAL_RESOURCE_ID

**Comment by tbieste** [comment]
Since this PR is dependent on PR #1930, waiting for that PR to be merged before marking this as ready for review.

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchTokenTest.java [hunk 1]
```diff
@@ -140,15 +160,6 @@ public void testSearchToken_code_chained_missing() throws Exception {
         assertSearchDoesntReturnComposition("subject:Basic.missing-code:missing", "false");
     }
 
-    @Test
-    public void testSearchToken_CodeableConcept_chained_missing() throws Exception {
```
Comment (lmsurpre) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchTokenTest.java:144]: any particular reason this one got removed?  or it was just redundant/unnecessary?

Comment (lmsurpre) [reply fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchTokenTest.java:144]: ah, never mind, just relocated.  now i see on line 239 of the new version

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 2]
```diff
@@ -1255,19 +1255,20 @@ private static void checkSearchParameterCombinations(Class<?> resourceType, List
...
                     // Optimization for search parameters that always reference the same system, added under #1929
-                    try {
-                        String implicitSystem = searchParameter.getExtension().stream()
-                                .filter(e -> SearchConstants.IMPLICIT_SYSTEM_EXT_URL.equals(e.getUrl()) && e.getValue() != null)
-                                .findFirst()
-                                .map(e -> e.getValue().as(Uri.class).getValue())
-                                .orElse(null);
-                        if (implicitSystem != null) {
-                            parameterValue.setValueSystem(implicitSystem);
+                    if (!Modifier.MISSING.equals(modifier)) {
```
Comment (lmsurpre) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1258]: ah, very interesting.  i definitely hadn't considered what might happen in here if there is a `:missing` modifier

**Review by lmsurpre — APPROVED** [review]
LGTM

**Review by michaelwschroeder — APPROVED** [review]
LGTM

**Comment by tbieste** [comment]
Ready to merge.

**Pull request merged**